presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -372,12 +372,13 @@ private void spillInput(Page page)
 
         long retainedSizeOfPage = page.getRetainedSizeInBytes();
         long sizeOfPage = page.getSizeInBytes();
-        log.debug("Spilling for operator %s, sizeOfPage %s, retinedSizeOfPage %s", operatorContext, sizeOfPage, retainedSizeOfPage);
+        long totalSpilledBytes = sizeOfPage + getSpiller().getSpilledPagesInMemorySize();
```

Comment:
won't this double count the page, as it should already be included in getSpilledPagesInMemorySize()?  is the problem that in some cases that's not updated yet by the time we get to this line?  Can we move the oom check to before we start spilling so we fail before we try spilling and also don't double count the page?

Reply:
Sorry my last changes were not published. Sure can move the check early so we can fail before spill

Reply:
Is getSpilledPagesInMemorySize() going to return correct accounting ? I see that spilling is done in separate thread and there is no construct to ensure memory visibility guarantee for spilledPagesInMemorySize variable. Am I missing something?

Reply:
That's a very good point.  and i guess we had the same issue before this change.  It's not a huge problem, as this is for failing fast before spill, but we'll fail regardless. However, probably we should make the fields volatile to ensure visibility.